### PR TITLE
Parquet reader: bug fix for a num_rows/skip_rows corner case, w/optimization for nested preprocessing

### PR DESCRIPTION
**Pull request by nvdbaranec**
Fixes an issue where using user bounds with parquet files containing both nested and non-nested types could result in incorrect row counts for the non-nested columns.  Originally reported by @etseidl 

The nature of the fix also implements a longstanding desired optimization:  when running the preprocess step for nested types, ignore pages for non-nested hierarchies.  This can result in significant speedups for files containing only a few nested columns.

<s>The tests added for this PR seem to tease a bug in the parquet writer into happening (https://github.com/rapidsai/cudf/issues/11748) so I will leave this as a draft until that issue is resolved.</s>

### CODE REVIEW

**Comment by ttnghia** [comment]
Wrongly switched branch. You may need to reverse changes from branch 22.12 and up merge with 22.10 instead.

**Comment by nvdbaranec** [comment]
rerun tests

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -1840,43 +1845,36 @@ void PreprocessColumnData(hostdevice_vector<PageInfo>& pages,
...
-                                    page_keys.end(),
-                                    size_input,
-                                    start_offset_output_iterator{pages.device_ptr(),
-                                                                 page_index.begin(),
-                                                                 0,
-                                                                 static_cast<int>(src_col_schema),
-                                                                 static_cast<int>(l_idx)});
+      // for nested hierarchies, compute per-page start offset.
+      // it would be better/safer to be checking (schema.max_repetition_level > 0) here, but there's
+      // no easy way to get at that info here. we'd have to move this function into reader_impl.cu
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_data.cu:1865]: Should there be an issue for this?

File: cpp/src/io/parquet/reader_impl.cu [hunk 2]
```diff
@@ -1353,26 +1353,39 @@ void reader::impl::preprocess_columns(hostdevice_vector<gpu::ColumnChunkDesc>& c
...
-    create_columns(_output_columns);
-  } else {
-    // preprocess per-nesting level sizes by page
+  // iterate over all input columns and allocate any associated output
+  // buffers if they are not part of a list hierarchy. mark down
+  // if we have any list columns that need further processing.
+  bool has_lists = false;
+  for (size_t idx = 0; idx < _input_columns.size(); idx++) {
+    auto const& input_col = _input_columns[idx];
+    size_t max_depth      = input_col.nesting_depth();
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/reader_impl.cu:1364]:
```suggestion
    size_t const max_depth      = input_col.nesting_depth();
```

File: cpp/src/io/parquet/reader_impl.cu [hunk 3]
```diff
@@ -1353,26 +1353,39 @@ void reader::impl::preprocess_columns(hostdevice_vector<gpu::ColumnChunkDesc>& c
...
+
+    auto* cols = &_output_columns;
+    for (size_t l_idx = 0; l_idx < input_col.nesting_depth(); l_idx++) {
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/reader_impl.cu:1367]:
```suggestion
    for (size_t l_idx = 0; l_idx < max_depth; l_idx++) {
```

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
I like how this is more understandable now. Thanks!

File: cpp/src/io/parquet/reader_impl.cu (hunk 2 above)
Comment (PointKernel) [inline cpp/src/io/parquet/reader_impl.cu:1364]:
```suggestion
    size_t const max_depth      = input_col.nesting_depth();
```

**Review by PointKernel — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by nvdbaranec** [comment]
@gpucibot merge

File: cpp/src/io/parquet/page_data.cu (hunk 1 above)
Comment (nvdbaranec) [reply cpp/src/io/parquet/page_data.cu:1865]: You were right.  This ended up being a bug :)  If you have a struct at the top of a nested hierarchy, this logic fails.  The `max_repetition_level` check is the correct one.

**Pull request merged**